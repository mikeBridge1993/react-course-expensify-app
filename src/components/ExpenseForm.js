import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';


export default class ExpenseForm extends React.Component {
  constructor (props){
    super(props);
    
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => {
      return {
        description
      }
    });
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => {
      return {
        note
      }
    });
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      console.log(amount)
      this.setState(() => {
        return {
          amount
        }
      });
    }
    
  };
  
  onDateChange = (createdAt) => {
   
    if (createdAt){
      this.setState(() => {
        return {
          createdAt
        }
      });
    }
    
  };

  onFocusedChange = ({ focused }) => {
    this.setState(() => {
      return {
        calendarFocused: focused
      }
    });
  };

  onSubmit = (e) => {
    e.preventDefault();
    console.log(this.state.createdAt.valueOf())
    if (!this.state.description || !this.state.amount){
      this.setState(() => {
      return {
        error: " Please provide description and amount."
      }
      });
    } else {
      console.log("great success");
      this.setState(() => {
        return {
          error: ''
        }
      });

      this.props.onSubmit(
        {
          description: this.state.description,
          amount: parseFloat(this.state.amount, 10)*100,
          createdAt: this.state.createdAt.valueOf(),
          note: this.state.note
        }
      );
    }
  };

  render() {
    return (
      <div className="container-fluid col-10 offset-1">
        {this.state.error && <div className="alert alert-danger mt-3 ">{this.state.error}</div>}
        <form onSubmit={this.onSubmit} className="p-2 mt-3 form">
          <div className="form-group">
            <input type="text" className="form-control" 
            autoFocus 
            placeholder="Description" 
            value={this.state.description}
            onChange={this.onDescriptionChange}
            />
          </div>
          <div className="form-group ">
            <input type="text" className="form-control" 
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
            />
          </div>
          <div className="form-group">
            <SingleDatePicker
              date={this.state.createdAt} // momentPropTypes.momentObj or null
              onDateChange={this.onDateChange} // PropTypes.func.isRequired
              focused={this.state.calendarFocused} // PropTypes.bool
              onFocusChange={this.onFocusedChange} // PropTypes.func.isRequired
              numberOfMonths={1}
              isOutsideRange={() => { false }}
              block
            />
          </div>
          <div className="form-group notes">
            <textarea type="text" className="form-control" placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
            ></textarea>
            
          </div>
          <button type="submit" className="btn btn-primary mt-2 btn-edit">{this.props.create ? "Create Expense" : "Edit Expense"}</button>
        </form>
      </div>
    )
  };
}

