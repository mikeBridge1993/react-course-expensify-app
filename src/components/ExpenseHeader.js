import React from 'react';

const ExpenseHeader = () => {
  return (
    <div className="container-fluid expenses">
      <hr className="my-0 py-2"/>
      <h1 className="display-4 my-0 py-3">Expenses List</h1>
      <hr className="mt-2 py-2"/>
    </div>
  );
}



export default ExpenseHeader;
