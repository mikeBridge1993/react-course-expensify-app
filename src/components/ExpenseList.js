import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


export const ExpenseList = (props) => {
  return (
    <div className="container-fluid col-10 offset-1">
      <ul className="list-group mb-2 pb-0 d-flex align-items-center">
        <li className="list-group-item rounded expense-list-header">
            <span className="mr-3">Expense</span>
            <span className="mr-3">Amount</span>
            <span className="mr-3">Date</span>
        </li>
      {props.expenses.map((el) => {
          return <ExpenseListItem key={el.id} {...el} createdAt={el.createdAt} />
      })}
      </ul>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
}

export default connect(mapStateToProps)(ExpenseList);

