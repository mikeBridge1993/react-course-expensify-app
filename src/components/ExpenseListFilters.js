import React from 'react';
import { connect } from 'react-redux';
import {setTextFilter ,sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate); 
    this.props.setEndDate(endDate); 
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => {
      return {
        calendarFocused
      };
    });
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  onSortChange = (e) => {
    console.log(e.target.value)
    if (e.target.value === "date-high"){
      this.props.sortByDate('high');
    } else if (e.target.value === "date-low") {
      this.props.sortByDate('low');
    } else if (e.target.value === "amount-high") {
      this.props.sortByAmount('high');
    } else if (e.target.value === "amount-low") {
      this.props.sortByAmount('low');
    }
  }

  render () {
    return (
      <div className="container-fluid expenses">
        <div className="input-group w-25">
          <input type="text" className="form-control" value={this.props.filters.text} onChange={this.onTextChange}/>
          <div className="input-group-append">
            <span className="input-group-text"><i className="fa fa-search"></i></span>
          </div>
        </div>
        <div className="w-25 my-2">
          <select className="form-control" 
            onChange={this.onSortChange}>
            <optgroup label="Date">            
              <option value="date-high">Newest first</option>
              <option value="date-low">Oldest first</option>
            </optgroup>
            <optgroup label="Amount">
              <option value="amount-high">Highest first</option>
              <option value="amount-low">Lowest first</option>
            </optgroup>
          </select>
        </div>
        <div className=" w-50 my-2">
            <DateRangePicker
              startDate={this.props.filters.startDate} // 
              startDateId="begginning" // PropTypes.string.isRequired,
              endDate={this.props.filters.endDate} // 
              endDateId="end" // PropTypes.string.isRequired,
              onDatesChange={this.onDatesChange} // 
              focusedInput={this.state.calendarFocused} // 
              onFocusChange={this.onFocusChange} //
              numberOfMonths={1} 
              showClearDates={true} 
              isOutsideRange={() => false} 
            />
          </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};


const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: criteria => dispatch(sortByAmount(criteria)),
    sortByDate: (criteria) => dispatch(sortByDate(criteria))
  };
};


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);