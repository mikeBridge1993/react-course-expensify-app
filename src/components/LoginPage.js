import React from 'react';
import { connect } from 'react-redux'
import { startLogin } from '../actions/auth'

export class LoginPage extends React.Component {
  
  render () {
    return (
      <div className="login-box d-flex align-items-center">
        <div className="login-box__box mx-auto bg-dark pb-3 col-4 rounded">
          <h1 className="login-box__title text-center pb-1 my-3">X<span className="text-light">pensify</span></h1>
          <p className="lead text-center text-secondary">It's time to get your expenses in check.</p>
          <button onClick={this.props.startLogin} className=" mx-auto text-center btn btn-lg btn-block col-8 offset-2 btn-login btn-primary mb-1 mt-4" type="submit">Sign in with <i className="fa fa-google text-danger"> </i> </button>
      </div> 
    </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
