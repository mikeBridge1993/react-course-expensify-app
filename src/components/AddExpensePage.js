import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startAddExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };

  render () {
    return (
      <div className="container-fluid  px-0">
        <div className="container-fluid d-flex justify-content-center p-4 summary">
        <p className="p-0 m-0 text-center">Create New Expense</p><br/>
        </div>   
        <ExpenseForm create={true} onSubmit={this.onSubmit}/>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);