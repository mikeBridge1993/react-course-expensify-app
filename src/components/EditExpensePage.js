import React from 'react';
import { connect } from 'react-redux';
import {startEditExpense, startRemoveExpense} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.startRemoveExpense({id: this.props.expense.id});
    this.props.history.push("/");
  };

  render () {
    return (
      <div className="container-fluid px-0">
        <div className="container-fluid d-flex justify-content-center p-4 summary">
        <p className="p-0 m-0 text-center">Edit Expense</p><br/>
        </div>   
        <ExpenseForm expense={this.props.expense} create={false} onSubmit={this.onSubmit}/>
        <div className="container-fluid col-10 offset-1">
          <button onClick={this.onRemove}
            className="btn ml-0 btn-danger mt-2 ml-2 btn-edit">
            Delete Expense</button> 
        </div>
      </div>
      
    );
  };
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
        return expense.id === props.match.params.id;
      }
    )
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);