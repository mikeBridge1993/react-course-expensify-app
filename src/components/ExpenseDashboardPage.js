import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseHeader from './ExpenseHeader';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
  <div className="container-fluid expenses px-0 bg-light">
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;