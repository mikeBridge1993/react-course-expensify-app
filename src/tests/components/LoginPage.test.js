import React from 'react'
import {shallow} from 'enzyme'
import {LoginPage} from '../../components/LoginPage'

test('should render Login Page correctly', () => {
  const wrapper = shallow(<LoginPage />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startlogin on button click', () => {
  const startLogin = jest.fn();
  const wrapper = shallow(<LoginPage startLogin={startLogin} />);

  wrapper.find('.btn-login').simulate('click');

  expect(startLogin).toHaveBeenLastCalledWith();

});