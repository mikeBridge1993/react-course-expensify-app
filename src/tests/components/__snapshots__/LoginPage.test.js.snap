// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login Page correctly 1`] = `
<div
  className="login-box d-flex align-items-center"
>
  <div
    className="login-box__box mx-auto bg-dark pb-3 col-4 rounded"
  >
    <h1
      className="login-box__title text-center pb-1 my-3"
    >
      X
      <span
        className="text-light"
      >
        pensify
      </span>
    </h1>
    <p
      className="lead text-center text-secondary"
    >
      It's time to get your expenses in check.
    </p>
    <button
      className=" mx-auto text-center btn btn-lg btn-block col-8 offset-2 btn-login btn-primary mb-1 mt-4"
      type="submit"
    >
      Sign in with 
      <i
        className="fa fa-google text-danger"
      >
         
      </i>
       
    </button>
  </div>
</div>
`;
